crates/telio-relay/src/multiplexer/inout.rs

```diff
@@ -25,21 +25,23 @@ where
     }
 
     /// Get references to split halves
-    pub fn split(&mut self) -> Option<(&mut SplitSink<T, I>, &mut SplitStream<T>)> {
+    pub fn split(&mut self) -> (&mut SplitSink<T, I>, &mut SplitStream<T>) {
         self.in_place(Self::into_split);
         match &mut self.0 {
-            Split(tx, rx) => Some((tx, rx)),
-            _ => None,
+            Split(tx, rx) => (tx, rx),
+            _ => unreachable!("After in_place(Self::into_split) self will always be split"),
         }
     }
 
     /// Get reference to joined type
-    pub fn joined(&mut self) -> Result<Option<&mut T>, ReuniteError<T, I>> {
-        self.try_in_place(Self::into_joined)?;
-        Ok(match &mut self.0 {
-            Joined(j) => Some(j),
-            _ => None,
-        })
+    pub fn joined(&mut self) -> &mut T {
+        self.try_in_place(Self::into_joined)
+            .expect("InOut always works with same underlying channel and it cannot fail on join");
+
+        match &mut self.0 {
+            Joined(j) => j,
+            _ => unreachable!("into_join enusures self will always be Joined"),
+        }
```

Comment:
I have interior mutability here, and it's hard to ensure this statically, this entire struct is basically for ensuring this in variance, so it would pretty for upper use in other area :/

Reply:
Now that I think about it, I'm worried that this is incorrect. For example in case of `fn split` the assumption expressed in the `unreachable!` is not correct - `self.0` can be still `Empty`.

Reply:
Well it's something that should not happen, the only time I tun self into empty wile using in_place.
There are no other ways to construct Empty, I think I will remove try join so it would not confuse this.

Reply:
I will queue property based test for this one :D

Reply:
LLT-4407 created for this